### ISSUE TITLE
Gateway server is failing Evaluate/Endorsement when non-endorsement org peers are bad

### ISSUE THREAD

**Issue by bviswana101** [hyperledger/fabric-gateway#557]
Channel default Endorsement Policy - Specific Org -  PeerOrg1
2 Peer Orgs (PeerOrg1, PeerOrg2) are joined as writers on the channel

Collection configuration includes both orgs

Chaincode A - does not use collection - is installed/approved on PeerOrg1 only.
Chaincode B - uses collection, is installed and approved on both PeerOrg1 and PeerOrg2.

Invocations (grpc.Method=Endorse) to Chaincode B and Chaincode A go to PeerOrg2 peers (as first endorser) even though its not part of endorsement policy.

**Comment by denyeart** [comment]
@bviswana101 Is your client targeting PeerOrg2 specifically? I believe gateway will attempt the first endorsement on the target peer, so it is up to the client to target at least one 'good' peer that it trusts, and then that peer can collect other endorsements as needed.

@bestbeforetoday keep me honest...

**Comment by bviswana101** [comment]
@denyeart No, infact we introduced a `ProposalOptions.endorsingOrg` to explicitly set it to PeerOrg1. 

Evaluate always fails when any peer in PeerOrg2 is not reachable.
Endorse intermittently fails when any peer in PeerOrg2 is not reachable.

I believe the Proposal is not sent to PeerOrg2 peers itself, However, it is not reasonable to expect that all peers are available at all times.

**Comment by denyeart** [comment]
@bviswana101  So you target PeerOrg1, the endorsement policy is PeerOrg1, and yet endorsement requests can fail if PeerOrg2 is not reachable by PeerOrg1?
If that is what you are saying, I agree that doesn't sound correct.

**Comment by bestbeforetoday** [comment]
I tried to reproduce bad peer/org targeting in isolation as follows:

1. Modified the **deployCC.sh** script in the Fabric [test-network](https://github.com/hyperledger/fabric-samples/tree/main/test-network) so that chaincode is installed to only the Org1 peer.
2. Brought up the test-network and deployed chaincode:
    ```bash
    ./network.sh up createChannel -c mychannel -ca
    ./network.sh deployCC -ccn basic -ccp ../asset-transfer-basic/chaincode-go/ -ccl go -ccep 'OR("Org1MSP.member","Org2MSP.member")'
    ```
3. Ran the **asset-transfer-basic/application-gateway-typescript** sample application, which submits and evaluates transactions using the Org1 Gateway peer and without any specific org targeting.

This worked fine.

I then repeated the test but this time installing the chaincode to only Org2 peer, while the application continues to use the Org1 Gateway peer. This also worked fine.

Perhaps the failures you are seeing are a special case of the behaviour reported in #556, where the Gateway is trying to establish connections to unreachable peers as part of its network discovery, not an issue of selecting incorrect endorsers.

**Comment by krishnanmano** [comment]
I have also come across a similar issue while using `fabric-gateway go client v1.3.0` but it got solved if I downgrade my client version to `fabric-gateway go client v1.2.2`. Please find my findings below.

Organizations: Org1, Org2, Org3
Peers: each organization has 3 peers among that 2 peers will be anchor peers.
Chaincode: Same version installed, approved, and committed in all organization's peers

Happy path:
Scenario 1: When all organization's anchor peers are up if org2 app invokes a smart contract on peer1Org2 that mutates state on private collection shared by Org1 and Org2 with endorsers as org1 and org2. Endorsement is successful and Block got committed.

Scenario 2: When all organization's anchor peers are up but org3's anchor peer is down(providing default TLS which is resulting in TLS Handshake error on other active peers). If the org2 app invokes a smart contract on peer1Org2 that mutates state on private collection shared by Org1 and Org2 with endorsers as org1 and org2. Endorsement is successful and Block got committed.

Failure Scenario: 
When all organization's anchor peers are up but Org3's anchor peer is down(providing default TLS which is resulting in TLS Handshake error on other active peers) and while Org3's peer was down if org2 peers got restarted. If the org2 app invokes a smart contract on peer1Org2 that mutates state on private collection shared by Org1 and Org2 with endorsers as org1 and org2, in this case, the endorsement is not successful. Even though the gateway client config timeout (as mentioned below). The application still waiting for Org3's peer to be up, once endorsement is successful and Block got committed.

```
gateway, err := client.Connect(
  id,
  client.WithSign(sign),
  client.WithClientConnection(grpcConn),
  client.WithEvaluateTimeout(5*time.Second),
  client.WithEndorseTimeout(15*time.Second),
  client.WithSubmitTimeout(5*time.Second),
  client.WithCommitStatusTimeout(1*time.Minute),
)
```

Issues:
1. Should not expect other Org's peers to be up and running as long as the Org is not endorsing the entity.
2. Gateway clients should honor the timeouts.

**Comment by bestbeforetoday** [comment]
@krishnanmano The first of the issues you report - Gateway peer waiting for org peers not required for endorsement - seems similar to the server-side mis-behaviour that this issue is addresses. This should not be affected by the version of the client API used. A fix went into Fabric v2.4.9 (and v2.5) that was supposed to address this server-side behaviour. More details in issue #556. Can you confirm which Fabric peer version you are using?

The Gateway client timeout behaviour seems to be something separate. The Fabric Gateway client API configures the default client-side timeout on the Context used for the underlying gRPC calls. There are unit tests that (should) confirm this happens, and I don't believe there has been any change to the code or behaviour in that area between [v1.2.2](https://github.com/hyperledger/fabric-gateway/releases/tag/v1.2.2) and[ v1.3.0](https://github.com/hyperledger/fabric-gateway/releases/tag/v1.3.0) of the client. What has changed is the version of the google.golang.org/grpc dependency. It is possible there was a bug or regression in behaviour of the gRPC version used by fabric-gateway [v1.3.0](https://github.com/hyperledger/fabric-gateway/releases/tag/v1.3.0). This dependency was updated again for the [v1.3.1](https://github.com/hyperledger/fabric-gateway/releases/tag/v1.3.1) release, so it might be that moving forward to that version also solves issues for you.

Which problem was solved for you by moving from fabric-gateway v1.3.0 back to v1.2.2?

**Comment by krishnanmano** [comment]
Thanks @bestbeforetoday for your reply. 
1. I am using the fabric peer version 2.4 now. 
2. By reverting gateway client from version 1.3.0 to 1.2.2, smart contact execution going through fine even though non endorsing org peers are down (returns tls handshake error), same is not working on 1.3.0.

As you suggested i will try the same scenarios with fabric peer version 2.4.9 or above and gateway client version 1.3.1.

**Comment by krishnanmano** [comment]
@bestbeforetoday, I have tried the same scenario with the latest version of gateway client 1.3.1 and Fabric peer version 2.4.9. Still, if Org3 peers are down and at the same time Org2 peers get restarted then the smart contract invoked on Org2 is failing, where Org3 is not endorser. Smart contract invokes working fine only after Org3 Peers come back online.

**Comment by bestbeforetoday** [comment]
Thanks for trying. I am still confused by the different behaviour you report depending on the client API version. All the client API does is send an endorse request to the peer Gateway service, and this has not changed at the client end. The peer Gateway service handles the endorsement (identifying the endorsement requirements, selecting endorsing peers, fail-over between peers, etc.).

Are you still seeing the problem that the client API is not timing out after the configured endorse timeout?

Looking back at your initial problem description, I have some questions to make sure I understand the scenario correctly:

1. You say, _"with endorsers as org1 and org2"_. Are you specifying the endorsing orgs explicitly at the client end, or letting the Gateway determine the required endorsers?
2. In the failure case:
    1. An org2 Gateway peer is used.
    2. Org 3 anchor peer is down.
    3. **And**, _"while Org3's peer was down if org2 peers got restarted"_
    
    Could you clarify the exact sequence of events for org2 and org3 peers that triggers the failure?
3. You say both, _"the endorsement is not successful"_, and, _"the application still waiting for Org3's peer to be up"_. Do you observe and endorsement failure, or do you just observe the endorsement process blocking until org3 peers are up?
5. Do you see evidence (perhaps from the Gateway peer logs) that it is waiting longer for peers to become available than the timeouts in the [Gateway peer configuration](https://github.com/hyperledger/fabric/blob/c2be53a22ab521380bf951bcb950fd7b466d7548/sampleconfig/core.yaml#L50-L61)?

Fabric v2.4 has been superseded by Fabric v2.5 LTS so investigation and any fixes will probably need to target that version. Is it possible to provide an absolutely minimal sample and scripts (or specific steps) than can be used to reproduce the failure?

**Comment by bestbeforetoday** [comment]
Closing due to inactivity - assuming resolved.
